File: ToDolist.java
import java.util.Scanner;

class Task {
    String task;
    int priority;
    boolean completed;
    Task next;

    Task(String task, int priority) {
        this.task = task;
        this.priority = priority;
        this.completed = false;
        this.next = null;
    }
}

public class ToDoList {
    private static Task head = null;
    private static Scanner sc = new Scanner(System.in);

    // Add a task
    public static void insertTask(String task, int priority) {
        if (task.isEmpty()) {
            System.out.println("\nTask cannot be empty!");
            return;
        }

        Task newTask = new Task(task, priority);

        if (head == null) {
            head = newTask;
        } else {
            Task temp = head;
            while (temp.next != null) {
                temp = temp.next;
            }
            temp.next = newTask;
        }
        System.out.println("\nTask added successfully.");
    }

    // Display tasks
    public static void displayTasks() {
        if (head == null) {
            System.out.println("\nNO TASKS AVAILABLE !!!");
            return;
        }
        Task temp = head;
        int i = 1;
        while (temp != null) {
            System.out.println(i + ". " + temp.task + " (Priority: " + temp.priority + ") [" +
                    (temp.completed ? "Completed" : "Not Completed") + "]");
            temp = temp.next;
            i++;
        }
    }

    // Delete task
    public static void deleteTask(String task) {
        if (head == null) {
            System.out.println("\nNO TASKS AVAILABLE !!!");
            return;
        }

        if (head.task.equals(task)) {
            head = head.next;
            System.out.println("\nTask deleted successfully.");
            return;
        }

        Task temp = head;
        Task prev = null;

        while (temp != null && !temp.task.equals(task)) {
            prev = temp;
            temp = temp.next;
        }

        if (temp == null) {
            System.out.println("\nTASK NOT FOUND !!!");
            return;
        }

        prev.next = temp.next;
        System.out.println("\nTask deleted successfully.");
    }

    // Mark task as completed
    public static void markCompleted(String task) {
        if (head == null) {
            System.out.println("\nNO TASKS AVAILABLE !!!");
            return;
        }

        Task temp = head;
        while (temp != null) {
            if (temp.task.equals(task)) {
                temp.completed = true;
                System.out.println("\nTask marked as completed.");
                return;
            }
            temp = temp.next;
        }
        System.out.println("\nTASK NOT FOUND !!!");
    }

    public static void main(String[] args) {
        int choice;
        String task;
        int priority;

        while (true) {
            System.out.println("\n<<<<<< TO-DO List >>>>>>");
            System.out.println("1. Add Task");
            System.out.println("2. Display Tasks");
            System.out.println("3. Delete Task");
            System.out.println("4. Mark Task as Completed");
            System.out.println("5. Exit");
            System.out.println("=======================================================");
            System.out.print("Enter your choice: ");

            choice = sc.nextInt();
            sc.nextLine(); // consume newline

            switch (choice) {
                case 1:
                    System.out.print("Enter the task: ");
                    task = sc.nextLine();
                    System.out.print("Enter priority: ");
                    priority = sc.nextInt();
                    sc.nextLine(); // consume newline
                    insertTask(task, priority);
                    break;

                case 2:
                    displayTasks();
                    break;

                case 3:
                    System.out.print("Enter the task to delete: ");
                    task = sc.nextLine();
                    deleteTask(task);
                    break;

                case 4:
                    System.out.print("Enter the task to mark as completed: ");
                    task = sc.nextLine();
                    markCompleted(task);
                    break;

                case 5:
                    System.out.println("Exiting the TO-DO LIST !!");
                    System.exit(0);

                default:
                    System.out.println("Invalid choice. Please try again.");
            }
        }
    }
}

File: TO-DO List.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

struct node {
    char task[100];
    int priority;              
    bool completed;             
    struct node* next;
};

struct node* head = NULL;

void trimNewline(char* str) {
    str[strcspn(str, "\n")] = '\0';
}


void insertTask(char task[], int priority) {
    if (strlen(task) == 0) {  
        printf("\nTask cannot be empty!\n");
        return;
    }

    struct node* newNode = (struct node*)malloc(sizeof(struct node));
    strcpy(newNode->task, task);
    newNode->priority = priority;
    newNode->completed = false;   
    newNode->next = NULL;

    if (head == NULL) {
        head = newNode;
    } else {
        struct node* temp = head;
        while (temp->next != NULL) {
            temp = temp->next;
        }
        temp->next = newNode;
    }
    printf("\nTask added successfully.\n");
}

void displayTasks() {
    int i = 1;
    if (head == NULL) {
        printf("\nNO TASKS AVAILABLE !!!\n");
    } else {
        struct node* ptr = head;
        while (ptr != NULL) {
            printf("%d. %s (Priority: %d) [%s]\n",
                   i, ptr->task, ptr->priority,
                   ptr->completed ? "Completed" : "Not Completed");
            i++;
            ptr = ptr->next;
        }
    }
}

void deleteTask(char task[]) {
    if (head == NULL) {
        printf("\nNO TASKS AVAILABLE !!!\n");
        return;
    }

    struct node* temp = head;
    struct node* prev = NULL;

    if (strcmp(temp->task, task) == 0) {
        head = temp->next;
        free(temp);
        printf("\nTask deleted successfully.\n");
        return;
    }

    while (temp != NULL && strcmp(temp->task, task) != 0) {
        prev = temp;
        temp = temp->next;
    }

    if (temp == NULL) {
        printf("\nTASK NOT FOUND !!!\n");
        return;
    }

    prev->next = temp->next;
    free(temp);
    printf("\nTask deleted successfully.\n");
}

void markCompleted(char task[]) {
    if (head == NULL) {
        printf("\nNO TASKS AVAILABLE !!!\n");
        return;
    }

    struct node* temp = head;
    while (temp != NULL) {
        if (strcmp(temp->task, task) == 0) {
            temp->completed = true;
            printf("\nTask marked as completed.\n");
            return;
        }
        temp = temp->next;
    }

    printf("\nTASK NOT FOUND !!!\n");
}

int getChoice() {
    char buffer[10];
    if (!fgets(buffer, sizeof(buffer), stdin)) return -1;
    trimNewline(buffer);

    for (int i = 0; buffer[i]; i++) {
        if (buffer[i] < '0' || buffer[i] > '9') {
            return -1; 
        }
    }
    return atoi(buffer);
}

int main() {
    int choice, priority;
    char task[100];

    while (true) {
        printf("\n<<<<<< TODO List >>>>>>\n");
        printf("1. Add Task\n");
        printf("2. Display Tasks\n");
        printf("3. Delete Task\n");
        printf("4. Mark Task as Completed\n");
        printf("5. Exit\n");
        printf("=======================================================\n");
        printf("Enter your choice: ");

        choice = getChoice();

        switch (choice) {
            case 1:
                printf("Enter the task: ");
                if (fgets(task, sizeof(task), stdin)) {
                    trimNewline(task);
                }
                printf("Enter priority: ");
                scanf("%d", &priority);
                getchar(); 
                insertTask(task, priority);
                break;

            case 2:
                displayTasks();
                break;

            case 3:
                printf("Enter the task to delete: ");
                if (fgets(task, sizeof(task), stdin)) {
                    trimNewline(task);
                }
                deleteTask(task);
                break;

            case 4:
                printf("Enter the task to mark as completed: ");
                if (fgets(task, sizeof(task), stdin)) {
                    trimNewline(task);
                }
                markCompleted(task);
                break;

            case 5:
                printf("Exiting the LIST !!\n");
                exit(0);

            default:
                printf("Invalid choice. Please try again.\n");
                break;
        }
    }
    return 0;
}

